File: packages/amplify_core/lib/src/types/models/auth_rule.dart
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

// ignore_for_file: constant_identifier_names

import 'dart:convert';

import 'package:collection/collection.dart';
import 'package:meta/meta.dart';

enum AuthStrategy { OWNER, GROUPS, PRIVATE, PUBLIC }

enum ModelOperation {
  CREATE,
  UPDATE,
  DELETE,
  READ,
  GET,
  LIST,
  SYNC,
  LISTEN,
  SEARCH
}

enum AuthRuleProvider { APIKEY, OIDC, IAM, USERPOOLS, FUNCTION }

@immutable
class AuthRule {
  const AuthRule({
    required this.authStrategy,
    this.ownerField,
    this.identityClaim,
    this.groupClaim,
    this.groups,
    this.groupsField,
    this.provider,
    this.operations,
  });

  final AuthStrategy authStrategy;
  final String? ownerField;
  final String? identityClaim;
  final String? groupClaim;
  final List<String>? groups;
  final String? groupsField;
  final AuthRuleProvider? provider;
  final List<ModelOperation>? operations;

  AuthRule copyWith({
    AuthStrategy? authStrategy,
    String? ownerField,
    String? identityClaim,
    String? groupClaim,
    List<String>? groups,
    String? groupsField,
    AuthRuleProvider? provider,
    List<ModelOperation>? operations,
  }) {
    return AuthRule(
      authStrategy: authStrategy ?? this.authStrategy,
      ownerField: ownerField ?? this.ownerField,
      identityClaim: identityClaim ?? this.identityClaim,
      groupClaim: groupClaim ?? this.groupClaim,
      groups: groups ?? this.groups,
      groupsField: groupsField ?? this.groupsField,
      provider: provider ?? this.provider,
      operations: operations ?? this.operations,
    );
  }

  Map<String, dynamic> toMap() {
    final map = {
      'authStrategy': authStrategy.name,
      'ownerField': ownerField,
      'identityClaim': identityClaim,
      'groupClaim': groupClaim,
      'groups': groups,
      'groupsField': groupsField,
      'provider': provider?.name,
      'operations': operations?.map((op) => op.name).toList(),
    };
    return Map<String, dynamic>.from(map)
      ..removeWhere((k, dynamic v) => v == null);
  }

  String toJson() => json.encode(toMap());

  @override
  String toString() {
    return 'AuthRule(authStrategy: $authStrategy, ownerField: $ownerField, identityClaim: $identityClaim, groupClaim: $groupClaim, groups: $groups, groupsField: $groupsField, provider: $provider, operations: $operations)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    final listEquals = const DeepCollectionEquality().equals;

    return other is AuthRule &&
        other.authStrategy == authStrategy &&
        other.ownerField == ownerField &&
        other.identityClaim == identityClaim &&
        other.groupClaim == groupClaim &&
        listEquals(other.groups, groups) &&
        other.groupsField == groupsField &&
        other.provider == provider &&
        listEquals(other.operations, operations);
  }

  @override
  int get hashCode {
    return authStrategy.hashCode ^
        ownerField.hashCode ^
        identityClaim.hashCode ^
        groupClaim.hashCode ^
        groups.hashCode ^
        groupsField.hashCode ^
        provider.hashCode ^
        operations.hashCode;
  }
}

File: packages/amplify_datastore/ios/Internal/Amplify/Categories/DataStore/Model/Internal/Schema/AuthRule.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

/// - Warning: Although this has `public` access, it is intended for internal use and should not be used directly
///   by host applications. The behavior of this may change without warning.
public enum AuthStrategy {
    case owner
    case groups
    case `private`
    case `public`
    case custom
}

/// - Warning: Although this has `public` access, it is intended for internal use and should not be used directly
///   by host applications. The behavior of this may change without warning.
public enum ModelOperation {
    case create
    case update
    case delete
    case read
}

/// - Warning: Although this has `public` access, it is intended for internal use and should not be used directly
///   by host applications. The behavior of this may change without warning.
public enum AuthRuleProvider {
    case apiKey
    case oidc
    case iam
    case userPools
    case function
}

/// - Warning: Although this has `public` access, it is intended for internal use and should not be used directly
///   by host applications. The behavior of this may change without warning.
public typealias AuthRules = [AuthRule]

/// - Warning: Although this has `public` access, it is intended for internal use and should not be used directly
///   by host applications. The behavior of this may change without warning.
public struct AuthRule {
    public let allow: AuthStrategy
    public let ownerField: String?
    public let identityClaim: String?
    public let groupClaim: String?
    public let groups: [String]
    public let groupsField: String?
    public let operations: [ModelOperation]
    public let provider: AuthRuleProvider?

    public init(allow: AuthStrategy,
                ownerField: String? = nil,
                identityClaim: String? = nil,
                groupClaim: String? = nil,
                groups: [String] = [],
                groupsField: String? = nil,
                provider: AuthRuleProvider? = nil,
                operations: [ModelOperation] = []) {
        self.allow = allow
        self.ownerField = ownerField
        self.identityClaim = identityClaim
        self.groupClaim = groupClaim
        self.groups = groups
        self.groupsField = groupsField
        self.provider = provider
        self.operations = operations
    }
}

extension AuthRule: Hashable {

}
